pubsub/jetstream/jetstream.go

```diff
@@ -88,8 +96,26 @@ func (js *jetstreamPubSub) Features() []pubsub.Feature {
 }
 
 func (js *jetstreamPubSub) Publish(req *pubsub.PublishRequest) error {
-	js.l.Debugf("Publishing topic %v with data: %v", req.Topic, req.Data)
-	_, err := js.jsc.Publish(req.Topic, req.Data)
+	var opts []nats.PubOpt
+	var msgID string
+
+	event, err := unmarshallCloudEvent(req)
+	if err != nil {
+		js.l.Debugf("error unmarshalling cloudevent: %v", err)
+	} else {
+		// Use the cloudevent id as the Nats-MsgId for deduplication
+		msgID = event.Context.GetID()
+		opts = append(opts, nats.MsgId(msgID))
+	}
+
+	if msgID == "" {
```

Comment:
It's nice to get the deduplication feature from Cloud Events, but it would be nice to check metadata for `Nats-Msg-Id` header (maybe first, before unmarshaling the Data?).

Reply:
Yes, I started there, and the metadata is empty. It is only populated by the developer who is publishing via that component.

Reply:
I think if we wanted to put `Nats-Msg-Id` into the metadata, we would need dapr/dapr to do that for us when the component is a pubsub.jetstream. It seemed more appropriate to put implementation detail into this repo though.

Reply:
A user of Dapr can optionally send custom metadata (think of it as a free form map / dictionary) with the publish request. This metadata is passed through to the component.

Most of the time a user will *not* be providing such metadata. PublishRequest metadata that is specific to a particular component is an anti pattern for the Dapr Runtime as it does not allow for an operator to simply replace the pubsub component without any code changes. But using such metadata for advanced scenarios may be acceptable.

Reply:
Hm, I get your point.

My point is: you might want to have NATS JetStream deduplication _without_ CloudEvents.

I'll get back to that tomorrow to see if there is a better way to achieve that.

Reply:
> My point is: you might want to have NATS JetStream deduplication without CloudEvents.

Correct, in which case my code wouldn't set the `Nats-Msg-Id`, and the developer will have to explicitly set `Nats-Msg-Id` in the metadata (which is then converted to Headers in #2201)

Reply:
@NickLarsenNZ yeah, agree. Let's have it like that.

I'll do metadata -> headers when this is merged and users will be able to do either - via CloudEvents or by putting the header.

Reply:
@Jarema Just chiming in, I noticed in the cloudevents/go-sdk repo, the `nats` and `nats_jetstream` support does not have the "binary content mode" implemented which relies on headers to decouple it from the payload. I had updated the NATS bindings spec [a while ago](https://github.com/cloudevents/spec/pull/1016), but never was actually implemented in the SDK, so I opened [this](https://github.com/cloudevents/sdk-go/issues/814). It seems like this would solve this problem at a lower-level since headers will be, properly, decoupled from the data.